vms/avm/service.go

```diff
@@ -231,6 +232,67 @@ func (service *Service) GetBalance(r *http.Request, args *GetBalanceArgs, reply
 	return nil
 }
 
+// GetAllBalancesArgs are arguments for calling into GetAllBalances
+type GetAllBalancesArgs struct {
+	Address string `json:"address"`
+}
+
+// GetAllBalancesReply is the response from a call to GetAllBalances
+type GetAllBalancesReply struct {
+	Balances map[string]json.Uint64 `json:"balances"`
+}
+
+// GetAllBalances returns a map where:
+//   Key: ID of an asset such that [args.Address] has a non-zero balance of the asset
+//   Value: The balance of the asset held by the address
+// Note that balances include assets that the address only _partially_ owns
+// (ie is one of several addresses specified in a multi-sig)
+func (service *Service) GetAllBalances(r *http.Request, args *GetAllBalancesArgs, reply *GetAllBalancesReply) error {
+	service.vm.ctx.Log.Verbo("GetAllBalances called with address: %s", args.Address)
+
+	address, err := service.vm.Parse(args.Address)
+	if err != nil {
+		return fmt.Errorf("couldn't parse given address: %s", err)
+	}
+	addrAsSet := ids.Set{}
+	addrAsSet.Add(ids.NewID(hashing.ComputeHash256Array(address)))
+
+	utxos, err := service.vm.GetUTXOs(addrAsSet)
+	if err != nil {
+		return fmt.Errorf("couldn't get address's UTXOs: %s", err)
+	}
+
+	assetIDs := ids.Set{}                    // IDs of assets the address has a non-zero balance of
+	balances := make(map[[32]byte]uint64, 0) // key: ID (as bytes). value: balance of that asset
+	for _, utxo := range utxos {
+		transferable, ok := utxo.Out.(ava.Transferable)
+		if !ok {
+			continue
+		}
+		assetID := utxo.AssetID()
+		assetIDs.Add(assetID)
+		if balance, ok := balances[assetID.Key()]; ok {
+			balance, err := safemath.Add64(transferable.Amount(), balance)
+			if err != nil {
+				balances[assetID.Key()] = math.MaxUint64
+			} else {
+				balances[assetID.Key()] = balance
+			}
+		} else {
+			balances[assetID.Key()] = transferable.Amount()
+		}
+	}
+
+	sortedAssetIDs := assetIDs.List() // sort so response is always in same order
+	ids.SortIDs(sortedAssetIDs)
+	reply.Balances = make(map[string]json.Uint64, len(sortedAssetIDs))
+	for _, assetID := range sortedAssetIDs {
+		reply.Balances[assetID.String()] = json.Uint64(balances[assetID.Key()])
+	}
```

Comment:
Maps aren't ordered in golang. Sorting the assetIDs here doesn't guarantee any ordering. Either we should just use `reply.Balances` above rather than making the `balances` map, or we should just iterate over the `balances` map here.